ffi/src/transaction/transaction_id.rs

```diff
@@ -0,0 +1,182 @@
+use crate::error::ExternResult;
+use crate::handle::Handle;
+use crate::transaction::ExclusiveTransaction;
+use crate::{
+    ExternEngine, IntoExternResult, KernelStringSlice, OptionalValue, SharedExternEngine,
+    SharedSnapshot, TryFromStringSlice,
+};
+use delta_kernel::transaction::Transaction;
+use delta_kernel::{DeltaResult, Snapshot};
+use std::sync::Arc;
+
+/// Associates an app_id and version with a transaction. These will be applied to the table on commit.
+///
+/// # Returns
+/// A new handle to the transaction that will set the `app_id` version to `version` on commit
+///
+/// # Safety
+/// Caller is responsible for passing [valid][Handle#Validity] handles. CONSUMES TRANSACTION
+#[no_mangle]
+pub unsafe extern "C" fn with_transaction_id(
+    txn: Handle<ExclusiveTransaction>,
+    app_id: KernelStringSlice,
+    version: i64,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveTransaction>> {
+    let txn = unsafe { txn.into_inner() };
+    let engine = unsafe { engine.as_ref() };
+    with_transaction_id_impl(*txn, app_id, version).into_extern_result(&engine)
+}
+
+fn with_transaction_id_impl(
+    txn: Transaction,
+    app_id: KernelStringSlice,
+    version: i64,
+) -> DeltaResult<Handle<ExclusiveTransaction>> {
+    let app_id_string: DeltaResult<String> = unsafe { TryFromStringSlice::try_from_slice(&app_id) };
+    Ok(Box::new(txn.with_transaction_id(app_id_string?, version)).into())
+}
+
+/// Retrieves the version associated with an app_id from a snapshot.
+///
+/// # Returns
+/// The version number if found, or an error of type `MissingDataError` when the app_id was not set
+///
+/// # Safety
+/// Caller must ensure [valid][Handle#Validity] handles are provided for snapshot and engine
+#[no_mangle]
+pub unsafe extern "C" fn get_app_id_version(
+    snapshot: Handle<SharedSnapshot>,
+    app_id: KernelStringSlice,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<OptionalValue<i64>> {
+    let snapshot = unsafe { snapshot.clone_as_arc() };
+    let engine = unsafe { engine.as_ref() };
+    let app_id = unsafe { String::try_from_slice(&app_id) };
+
+    get_app_id_version_impl(snapshot, app_id, engine)
+        .map(OptionalValue::from)
+        .into_extern_result(&engine)
+}
+
+fn get_app_id_version_impl(
+    snapshot: Arc<Snapshot>,
+    app_id: DeltaResult<String>,
+    extern_engine: &dyn ExternEngine,
+) -> DeltaResult<Option<i64>> {
+    snapshot.get_app_id_version(&app_id?, extern_engine.engine().as_ref())
+}
```

Comment:
do we need this or could we just inline `snapshot.get_app_id_version(...)` above?

Reply:
well the app_id might be a `DeltaResult::Err` right? so then we'd need to handle that anyway so I thought this was simpler and more in line with the other ffi methods?

Perhaps my rust skills are lacking here, but how else would we cleanly return an `ExternResult::Err` from `get_app_id_version` if `app_id` is a `DeltaResult::Err`?